### PR TITLE
fix(transaction): Don't set continuation for blocking

### PR DESCRIPTION
**Pull request by dranikpg**
Cause: `Check failed: continuation_trans_ == nullptr || continuation_trans_ == trans BLMOVE@4135983/2 {id=33648} when polling BLMOVE@4135965/1 {id=50088}cont_mask: 41 vs 41`

Preconditions: 
* Two BLMOVE commands waiting for different keys on _the same shard_, for both of them the source and destination key are located on different shards, so they need _two hops_ to get the job done
* _A single_ MULTI/EXEC unblocking both of them

Flow:
* Both suspended BLMOVEs wake up and continue running PollExecution
* Because they are 2-hop commands, they try to set continuation_trans_ 
* The first one does so successfully, the second doesn't expect it to be set, so it fails on the check above

Fix:
* Don't set continuation_trans_ for awoken blocking commands that want to proceed with another hop
* Is it safe in regard to other commands? Yes, because we don't continue execution on the transaction queue while there are awakened transactions
* We remove ourselfs from the awakened transaction set only when we conclude
* We unlock the keys only when we conclude, so OOO is not possible (ooo doesn't check for awakened)

Follow up:
* Rethink and improve blocking flow in future code

### CODE REVIEW

File: src/server/engine_shard.cc [hunk 1]
```diff
@@ -486,14 +486,16 @@ void EngineShard::PollExecution(const char* context, Transaction* trans) {
...
+    DCHECK(trans->GetNamespace().GetBlockingController(shard_id_)->HasAwakedTransaction());
     CHECK(continuation_trans_ == nullptr || continuation_trans_ == trans)
         << continuation_trans_->DebugId() << " when polling " << trans->DebugId()
         << "cont_mask: " << continuation_trans_->DEBUG_GetLocalMask(sid) << " vs "
         << trans->DEBUG_GetLocalMask(sid);
 
     // Commands like BRPOPLPUSH don't conclude immediately
     if (trans->RunInShard(this, false)) {
-      continuation_trans_ = trans;
+      // execution is blocked while the awakened set set is non-empty, so no need to set
```
Comment (romange) [inline src/server/engine_shard.cc:497]:
```suggestion
      // execution is blocked while HasAwakedTransaction() returns true, so no need to set continuation_trans_. 
     // Moreover, setting it for wakened multi-hop transactions may lead to inconcistency, see BLMove22 test.
```

File: src/server/engine_shard.cc [hunk 2]
```diff
@@ -486,14 +486,16 @@ void EngineShard::PollExecution(const char* context, Transaction* trans) {
     return;
 
   if (trans_mask & Transaction::AWAKED_Q) {
+    DCHECK(trans->GetNamespace().GetBlockingController(shard_id_)->HasAwakedTransaction());
     CHECK(continuation_trans_ == nullptr || continuation_trans_ == trans)
```
Comment (romange) [inline src/server/engine_shard.cc:490]: please update the CHECK here as well.

**Review by romange — COMMENTED** [review]
LGTM

File: src/server/engine_shard.cc (hunk 2 above)
Comment (dranikpg) [reply src/server/engine_shard.cc]: Done

File: src/server/engine_shard.cc [hunk 3]
```diff
@@ -486,14 +486,18 @@ void EngineShard::PollExecution(const char* context, Transaction* trans) {
...
         << "cont_mask: " << continuation_trans_->DEBUG_GetLocalMask(sid) << " vs "
         << trans->DEBUG_GetLocalMask(sid);
 
     // Commands like BRPOPLPUSH don't conclude immediately
     if (trans->RunInShard(this, false)) {
-      continuation_trans_ = trans;
+      // execution is blocked while HasAwakedTransaction() returns true, so no need to set
+      // continuation_trans_. Moreover, setting it for wakened multi-hop transactions may lead to
+      // inconcistency, see BLMoveSimultaneously test.
+      // continuation continuation_trans_ = trans;
```
Comment (romange) [inline src/server/engine_shard.cc:500]: this line not needed, or remove `continuation `

**Review by romange — DISMISSED** [review]
LGTM

**Comment by dranikpg** [comment]
Updated PR description

**Pull request merged**